File: Samples/iOS/Objective-C/IOS-SDK-Demo-App/Ad Formats/LOInlineViewController.m
//
//  LOInlineViewController.m
//  test-objc-fw-integration
//

#import "LOInlineViewController.h"
#import <VungleAdsSDK/VungleAdsSDK.h>

@interface LOInlineViewController () <VungleBannerViewDelegate, UITableViewDelegate, UITableViewDataSource>
@property (nonatomic, strong) NSMutableArray *models;
@property (nonatomic, strong) UITableViewCell *adCell;
@property (nonatomic, strong) VungleBannerView *adView;
@property (nonatomic, weak) IBOutlet UITableView *tableView;
@end

@implementation LOInlineViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.adView = [[VungleBannerView alloc] initWithPlacementId:self.placementId 
                                                   vungleAdSize:[VungleAdSize VungleAdSizeFromCGSize:CGSizeMake(300, 300)]];
    self.adView.delegate = self;
    [self.adView load:nil];
    self.models = [[NSMutableArray alloc] init];
    for (int i = 1; i <= 10; i++)
    {
        [self.models addObject: [InlineModel new]];
    }
    self.adCell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier: @"adCell"];
    self.navigationItem.title = @"Liftoff Inline Ad";
}

#pragma mark - UITableViewDelegate / UITableViewDatasource methods

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.models.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    InlineModel *model = self.models[indexPath.row];
    if (model.isAd) {
        return self.adCell;
    }
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"InlineCell"
                                                                     forIndexPath:indexPath];
    [cell.imageView setImage: [UIImage imageNamed:@"Liftoff-monetize-demo"]];
    cell.textLabel.text = @"Inline demo";
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    InlineModel *model = self.models[indexPath.row];
    return model.cellHeight;
}

#pragma mark - VungleInterstitial Delegate Methods
- (void)bannerAdDidLoad:(VungleBannerView * _Nonnull)bannerView {
    NSLog(@"bannerAdDidLoad");
    InlineModel *model = [[InlineModel alloc] init];
    model.isAd = YES;
    model.cellHeight = 300;
    [self.adCell.contentView addSubview: bannerView];
    [self.models insertObject: model atIndex:2];
    [self.tableView reloadData];
}

- (void)bannerAdWillPresent:(VungleBannerView * _Nonnull)bannerView {
    NSLog(@"bannerAdWillPresent");
}

- (void)bannerAdDidPresent:(VungleBannerView * _Nonnull)bannerView {
    NSLog(@"bannerAdDidPresent");
}

- (void)bannerAdDidFail:(VungleBannerView * _Nonnull)bannerView withError:(NSError * _Nonnull)withError {
    NSLog(@"bannerAdDidFail, %@", [withError localizedDescription]);
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Load Ad Error"
                                                                   message: [withError localizedDescription]
                                                            preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:@"Dismiss" style:UIAlertActionStyleDefault handler:nil]];
    [self presentViewController:alert animated:YES completion:nil];
}
- (void)bannerAdWillClose:(VungleBannerView * _Nonnull)bannerView {
    NSLog(@"bannerAdWillClose");
}

- (void)bannerAdDidClose:(VungleBannerView * _Nonnull)bannerView {
    NSLog(@"bannerAdDidClose");
    [self.models removeObjectAtIndex:2];
    [self.tableView reloadData];
}

- (void)bannerAdDidTrackImpression:(VungleBannerView * _Nonnull)bannerView {
    NSLog(@"bannerAdDidTrackImpression");
}

- (void)bannerAdDidClick:(VungleBannerView * _Nonnull)bannerView {
    NSLog(@"bannerAdDidClick");
}

- (void)bannerAdWillLeaveApplication:(VungleBannerView * _Nonnull)bannerView {
    NSLog(@"bannerAdWillLeaveApplication");
}
@end


@implementation InlineModel
- (instancetype)init
{
    self = [super init];
    if (self) {
        self.isAd = NO;
        self.cellHeight = 200;
    }
    return self;
}

@end

File: Samples/iOS/Swift/IOS-SDK-Demo-App/Ad Formats/LOInlineViewController.swift
//
//  LOInlineViewController.swift
//  PublicSampleAppTrial
//

import UIKit
import VungleAdsSDK

class LOInlineViewController: UIViewController {

    @IBOutlet weak var tableView: UITableView!

    var placementId: String?
    var adView: VungleBannerView?
    var models = [InlineModel].init(repeating: InlineModel(), count: 10)
    let adCell = InlineAdCell(style: .default, reuseIdentifier: "adCell")

    override func viewDidLoad() {
        super.viewDidLoad()
        navigationItem.title = "Liftoff Inline Ad"
        if let placementId = placementId {
            adView = VungleBannerView(placementId: placementId, vungleAdSize: .VungleAdSizeFromCGSize(CGSize(width: 300, height: 300)))
            adView?.delegate = self
            adView?.load()
        }
    }

    @IBAction func btnPressed(_ sender:UIButton) {
    }
}

extension LOInlineViewController: UITableViewDelegate, UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return models.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let model = models[indexPath.row]
        if model.isAd {
            return adCell
        }
        guard let cell = tableView.dequeueReusableCell(withIdentifier: "InlineCell", for: indexPath) as? InlineCell else {
            return UITableViewCell()
        }
        return cell
    }

    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return models[indexPath.row].cellHeight
    }
}

extension LOInlineViewController: VungleBannerViewDelegate {
    func bannerAdDidLoad(_ bannerView: VungleAdsSDK.VungleBannerView) {
        var adModel = InlineModel()
        adModel.isAd = true
        adModel.cellHeight = 300
        models.insert(adModel, at: 1)
        adCell.contentView.addSubview(bannerView)
        tableView.reloadData()
    }

    func bannerAdWillPresent(_ bannerView: VungleAdsSDK.VungleBannerView) {
        print("bannerAdWillPresent")
    }

    func bannerAdDidPresent(_ bannerView: VungleAdsSDK.VungleBannerView) {
        print("bannerAdDidPresent")
    }

    func bannerAdDidFail(_ bannerView: VungleAdsSDK.VungleBannerView, withError: NSError) {
        let alert = UIAlertController(title: "Load Ad Error", message: withError.localizedDescription, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "Dismiss", style: .default))
        present(alert, animated: true,completion: nil)
    }

    func bannerAdWillClose(_ bannerView: VungleAdsSDK.VungleBannerView) {
        print("bannerAdWillClose")
    }

    func bannerAdDidClose(_ bannerView: VungleAdsSDK.VungleBannerView) {
        print("bannerAdDidClose")
        models = models.filter({ !$0.isAd })
        tableView.reloadData()
    }

    func bannerAdDidTrackImpression(_ bannerView: VungleAdsSDK.VungleBannerView) {
        print("bannerAdDidTrackImpression")
    }

    func bannerAdDidClick(_ bannerView: VungleAdsSDK.VungleBannerView) {
        print("bannerAdDidClick")
    }

    func bannerAdWillLeaveApplication(_ bannerView: VungleAdsSDK.VungleBannerView) {
        print("bannerAdWillLeaveApplication")
    }
}

struct InlineModel {
    var cellHeight: CGFloat = 200
    var isAd = false
}
